book/src/examples/calculator.md

```diff
@@ -0,0 +1,139 @@
+# Simple calculator
+
+When you implement an interpreter for a DSL or a programming language, the process typically involves the following steps:
+
+1. **Lexing**: Splitting the input into tokens via a lexer.
+
+2. **Parsing**: Converting the tokens into an [AST (abstract syntax tree)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) via a parser.
+
+3. **Evaluation**: Evaluating the AST to produce the result.
+
+In this example, we implement a simple calculator that evaluates arithmetic expressions such as `1 + 2 * 3` or `((1 + 2) * 3 + 4) * 2 + 4 / 3`.
+
+We use `logos` as the lexer generator and [`chumsky`](https://github.com/zesterer/chumsky) as the parser generator.
+
+![flow chart](/assets/calculator_example_flow.png)
+
+## 1. Try It
+
+Before diving into the implementation details, let's play with it.
+
+```bash
+$ cargo run --example calculator '1 + 7 * (3 - 4) / 2'
+```
+
+**Output**:
+
+```
+[AST]
+Add(
+    Int(
+        1,
+    ),
+    Div(
+        Mul(
+            Int(
+                7,
+            ),
+            Sub(
+                Int(
+                    3,
+                ),
+                Int(
+                    4,
+                ),
+            ),
+        ),
+        Int(
+            2,
+        ),
+    ),
+)
+
+[result]
+-2
+```
+
+~~~admonish note title="Full Code" collapsible=true
+
+```rust,no_run,noplayground
+{{#include ../../../examples/calculator.rs:all}}
+```
+
+~~~
+
+## 2. Lexer
+
+Our calculator supports the following tokens:
+
+- Integer literals: `0`, `1`, `15`, etc.
+
+- Unary operator: `-`
+
+- Binary operators: `+`, `-`, `*`, `/`
+
+- Parenthesized expressions: `(3 + 5) * 2`, `((1 + 2) * 3 + 4) * 2 + 3 / 2`, etc.
+
+```rust,no_run,noplayground
+{{#include ../../../examples/calculator.rs:tokens}}
+```
+
+## 3. Parser
+
+While it is easy enough to manually implement a parser in this case (e.g. [Pratt parsing](https://en.wikipedia.org/wiki/Operator-precedence_parser#Pratt_parsing)), let's just use [`chumsky`](https://github.com/zesterer/chumsky) crate, which is one of the most popular parser generator libraries in Rust.
+
+### 3.1 AST Definition
+
+First, we define the AST.
+
+```rust,no_run,noplayground
+{{#include ../../../examples/calculator.rs:ast}}
+```
+
+Note that
+
+- We name the enum not `AST` but `Expr` because an AST is just nested expressions.
+
+- There is no `Parenthesized` variant because parentheses only affect the order of operations (i.e. precedence), which is reflected in the AST structure.
```

Comment:
```suggestion
- There is no `Parenthesized` variant because parentheses only affect the order of operations (i.e., precedence), which is reflected in the AST structure.
```